contracts/main/CurveCryptoMathOptimized3.vy

```diff
@@ -552,220 +550,55 @@ def get_p(
-            unsafe_div(s3 * x1, D) * x2, D
-        ) * x3,
-        D
-    )
+    # constant which is in 10**36 precision:
+    NN_A_gamma2: int256 = unsafe_mul(27 * ANN, unsafe_mul(gamma, gamma))
 
-    # 10**18*19683*x1/D*x1/D*x1/D*x2/D*x2/D*x2/D*x3/D*x3/D*x3/D
-    # Same as before.
-    s4: int256 = 10**18 * 19683 * x1
-    s4 = unsafe_div(unsafe_div(s4, D) * x2, D) * x3
-    s4 = unsafe_div(
+    # K = P * N**N / D**N.
+    # K is dimensionless and has 10**36 precision:
+    K: int256 = unsafe_div(
         unsafe_div(
-            unsafe_div(
-                unsafe_div(
-                    unsafe_div(
-                        unsafe_div(
-                            unsafe_div(s4, D) * x1,
-                            D
-                        ) * x2,
-                        D
-                    ) * x3,
-                    D,
-                ) * x1,
-                D
-            ) * x2,
+            unsafe_div(27 * x * y, D) * z,
             D
-        ) * x3,
+        ) * 10**36,
         D
     )
 
-    a: int256 = s1 + s2 + s4 - s3
-
-    # 10**18*729*A*x1/D*x2/D*x3/D*gamma**2/D/27/10000
-    b: int256 = unsafe_div(
-        unsafe_div(
-            unsafe_div(
-                unsafe_div(
-                    unsafe_div(
-                        unsafe_mul(10**18 * 27, ANN),
-                        convert(A_MULTIPLIER, int256),
-                    ) * x1,
-                    D,
-                ) * x2,
-                D,
-            ) * x3,
-            D,
-        ) * gamma2,
-        D,
+    # G = (
+    #     3 * K**2
+    #     - 2 * K * (2 * gamma + 3)
+    #     + N_COINS**N_COINS * A * gamma**2 * (S - D) / D
+    #     + (gamma + 1) * (gamma + 3)
+    # )
+    # G is in 10**36 precision and is dimensionless.
+    G: int256 = (
+        unsafe_div(3 * K**2, 10**36)
+        - unsafe_div(K * (unsafe_add(unsafe_mul(4 * gamma, 10**18), 6*10**36)), 10**36)
+        + unsafe_div(unsafe_div(unsafe_div(NN_A_gamma2 * (S - D), D), 27), 10000)
+        + unsafe_mul(unsafe_add(gamma, 10**18), unsafe_add(gamma, 3*10**18))
     )
 
-    # 27*A*gamma**2*(10**18 + gamma)/D/27/10000
-    # entire expression can be unsafe:
-    c: int256 = unsafe_div(
-        unsafe_div(
-            unsafe_div(
-                unsafe_mul(
-                    27,
-                    unsafe_mul(
-                        ANN,
-                        unsafe_mul(
-                            gamma2,
-                            unsafe_add(10**18, gamma)
-                        )
-                    )
-                ),
-                D
-            ),
-            27
-        ),
-        convert(A_MULTIPLIER, int256)
-    )
+    # Safety check: ensure G is positive:
+    assert G > 0, "dev: G must be greater than 0"
 
+    # G3 = G * D / (N**N * ANN * gamma**2)
+    # G3 is dimensionless. It is not in 10**36 precision (lower, about 10**20):
+    G3: uint256 = convert(unsafe_div(G * D * 27 * 10000, NN_A_gamma2), uint256)
```

Comment:
and here finalized G3 would be `convert(unsafe_div(G * D, NN_A_gamma2), uint256)`, it is proportional to D - dollar value, multiplied by dimensionless parameter